cartographer/mapping_3d/sparse_pose_graph/optimization_problem.cc

```diff
@@ -259,6 +259,55 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
     }
   }
 
+  // Add fixed frame pose constraints.
+  std::deque<CeresPose> C_fixed_frames;
+  for (size_t trajectory_id = 0; trajectory_id != node_data_.size();
+       ++trajectory_id) {
+    if (trajectory_id >= fixed_frame_pose_data_.size()) {
+      break;
+    }
+
+    bool fixed_frame_pose_initialized = false;
+
+    const auto& node_data = node_data_[trajectory_id];
+    for (size_t node_index = 0; node_index < node_data.size(); ++node_index) {
+      if (!fixed_frame_pose_data_.at(trajectory_id)
+               .Has(node_data[node_index].time)) {
+        continue;
+      }
+
+      const mapping::SparsePoseGraph::Constraint::Pose constraint_pose{
+          fixed_frame_pose_data_.at(trajectory_id)
+              .Lookup(node_data[node_index].time),
+          options_.fixed_frame_pose_translation_weight(),
+          options_.fixed_frame_pose_rotation_weight()};
+
+      // TODO(zhengj, whess): Allow choosing rotation parameterization(supports
+      // non-gravity-aligned fixed frames)
+      if (!fixed_frame_pose_initialized) {
+        C_fixed_frames.emplace_back(
+            transform::Rigid3d(
+                constraint_pose.zbar_ij.translation(),
```

Comment:
This should use an initial estimate of the fixed frame pose in place of `constraint_pose.zbar_ij`.
`constraint_pose.zbar_ij` is from node to fixed frame and the pose in `node_data_` from node to map frame and we want the fixed frame in the map frame: `node_data_[...]... * constraint_pose.zbar_ij.inverse()`?

Reply:
I see, thanks!